File: backend/schemas/fir.py
from datetime import date, datetime
from pydantic import BaseModel, Field
from typing import List, Optional

# ==========================================
# Lookup / Master Table DTOs
# ==========================================

class LookupDTO(BaseModel):
    id: int
    name: str
    active: bool = True
    sort_order: int = 0

    model_config = {
        "from_attributes": True
    }

class StateDTO(LookupDTO):
    NationalityID: Optional[int] = None

class DistrictDTO(LookupDTO):
    StateID: int

class CourtDTO(LookupDTO):
    DistrictID: int
    StateID: int

class UnitTypeDTO(LookupDTO):
    CityDistState: Optional[str] = None
    Hierarchy: Optional[int] = None

class UnitDTO(LookupDTO):
    TypeID: int
    ParentUnit: Optional[int] = None
    NationalityID: Optional[int] = None
    StateID: int
    DistrictID: int

class RankDTO(LookupDTO):
    Hierarchy: Optional[int] = None

class DesignationDTO(LookupDTO):
    SortOrder: Optional[int] = None

class EmployeeDTO(BaseModel):
    id: int
    DistrictID: int
    UnitID: int
    RankID: int
    DesignationID: int
    KGID: str
    FirstName: str
    EmployeeDOB: Optional[date] = None
    GenderID: int
    BloodGroupID: Optional[int] = None
    PhysicallyChallenged: bool = False
    AppointmentDate: Optional[date] = None
    active: bool = True

    model_config = {
        "from_attributes": True
    }

class CaseCategoryDTO(LookupDTO):
    pass

class GravityOffenceDTO(LookupDTO):
    pass

class CaseStatusDTO(LookupDTO):
    pass

class CasteDTO(LookupDTO):
    pass

class ReligionDTO(LookupDTO):
    pass

class OccupationDTO(LookupDTO):
    pass

class GenderDTO(LookupDTO):
    pass

class NationalityDTO(LookupDTO):
    pass

class BloodGroupDTO(LookupDTO):
    pass

class ActDTO(BaseModel):
    ActCode: str
    ActDescription: Optional[str] = None
    ShortName: Optional[str] = None
    active: bool = True

    model_config = {
        "from_attributes": True
    }

class SectionDTO(BaseModel):
    ActCode: str
    SectionCode: str
    SectionDescription: Optional[str] = None
    active: bool = True

    model_config = {
        "from_attributes": True
    }

class CrimeHeadDTO(BaseModel):
    id: int
    CrimeGroupName: str
    active: bool = True

    model_config = {
        "from_attributes": True
    }

class CrimeSubHeadDTO(BaseModel):
    id: int
    CrimeHeadID: int
    CrimeHeadName: str
    SeqID: Optional[int] = None

    model_config = {
        "from_attributes": True
    }

# ==========================================
# Transactional Incident DTOs
# ==========================================

class InvOccuranceTimeBase(BaseModel):
    IncidentFromDate: Optional[datetime] = None
    IncidentToDate: Optional[datetime] = None
    InfoReceivedPSDate: Optional[datetime] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    BriefFacts: Optional[str] = None

class InvOccuranceTimeCreate(InvOccuranceTimeBase):
    pass

class InvOccuranceTimeResponse(InvOccuranceTimeBase):
    CaseMasterID: int
    created_at: datetime
    updated_at: datetime

    model_config = {
        "from_attributes": True
    }

class ComplainantDetailsBase(BaseModel):
    ComplainantName: str
    AgeYear: Optional[int] = None
    OccupationID: Optional[int] = None
    ReligionID: Optional[int] = None
    CasteID: Optional[int] = None
    GenderID: Optional[int] = None

class ComplainantDetailsCreate(ComplainantDetailsBase):
    pass

class ComplainantDetailsResponse(ComplainantDetailsBase):
    id: int
    CaseMasterID: int
    created_at: datetime
    updated_at: datetime

    model_config = {
        "from_attributes": True
    }

class FIRVictimBase(BaseModel):
    VictimName: str
    AgeYear: Optional[int] = None
    GenderID: Optional[int] = None
    VictimPolice: str = "0"

class FIRVictimCreate(FIRVictimBase):
    pass

class FIRVictimResponse(FIRVictimBase):
    id: int
    CaseMasterID: int
    created_at: datetime
    updated_at: datetime

    model_config = {
        "from_attributes": True
    }

class AccusedBase(BaseModel):
    AccusedName: str
    AgeYear: Optional[int] = None
    GenderID: Optional[int] = None
    PersonID: Optional[str] = None

class AccusedCreate(AccusedBase):
    pass

class AccusedResponse(AccusedBase):
    id: int
    CaseMasterID: int
    created_at: datetime
    updated_at: datetime

    model_config = {
        "from_attributes": True
    }

class ActSectionAssociationBase(BaseModel):
    ActCode: str
    SectionCode: str
    ActOrderID: Optional[int] = None
    SectionOrderID: Optional[int] = None

class ActSectionAssociationCreate(ActSectionAssociationBase):
    pass

class ActSectionAssociationResponse(ActSectionAssociationBase):
    CaseMasterID: int
    created_at: datetime
    updated_at: datetime

    model_config = {
        "from_attributes": True
    }

class ArrestSurrenderBase(BaseModel):
    ArrestSurrenderTypeID: int
    ArrestSurrenderDate: date
    ArrestSurrenderStateId: int
    ArrestSurrenderDistrictId: int
    PoliceStationID: int
    IOID: int
    CourtID: int
    AccusedMasterID: int
    IsAccused: bool = True
    IsComplainantAccused: bool = False

class ArrestSurrenderCreate(ArrestSurrenderBase):
    associated_accused_ids: List[int] = []

class ArrestSurrenderResponse(ArrestSurrenderBase):
    id: int
    CaseMasterID: int
    created_at: datetime
    updated_at: datetime
    associated_accused_ids: List[int] = []

    model_config = {
        "from_attributes": True
    }

class ChargesheetDetailsBase(BaseModel):
    csdate: datetime
    cstype: str
    PolicePersonID: int

class ChargesheetDetailsCreate(ChargesheetDetailsBase):
    pass

class ChargesheetDetailsResponse(ChargesheetDetailsBase):
    id: int
    CaseMasterID: int
    created_at: datetime
    updated_at: datetime

    model_config = {
        "from_attributes": True
    }

class CaseMasterBase(BaseModel):
    CrimeNo: Optional[str] = None
    CaseNo: Optional[str] = None
    CrimeRegisteredDate: Optional[date] = None
    PolicePersonID: int
    PoliceStationID: int
    CaseCategoryID: int
    GravityOffenceID: int
    CrimeMajorHeadID: int
    CrimeMinorHeadID: int
    CaseStatusID: int
    CourtID: int
    BriefFacts: Optional[str] = None
    dataset_id: Optional[int] = None

class CaseMasterCreate(CaseMasterBase):
    occurrence_time: Optional[InvOccuranceTimeCreate] = None
    complainants: List[ComplainantDetailsCreate] = []
    victims: List[FIRVictimCreate] = []
    accused: List[AccusedCreate] = []
    act_sections: List[ActSectionAssociationCreate] = []
    arrest_surrenders: List[ArrestSurrenderCreate] = []
    chargesheets: List[ChargesheetDetailsCreate] = []

class CaseMasterResponse(CaseMasterBase):
    id: int
    created_at: datetime
    updated_at: datetime

    occurrence_time: Optional[InvOccuranceTimeResponse] = None
    complainants: List[ComplainantDetailsResponse] = []
    victims: List[FIRVictimResponse] = []
    accused: List[AccusedResponse] = []
    act_sections: List[ActSectionAssociationResponse] = []
    arrest_surrenders: List[ArrestSurrenderResponse] = []
    chargesheets: List[ChargesheetDetailsResponse] = []

    model_config = {
        "from_attributes": True
    }

File: frontend/features/fir/types/fir.ts
// ══════════════════════════════════════════════════════════════════════════════
// FIR Case Management — TypeScript Type Definitions
// Mirrors backend/schemas/fir.py Pydantic models exactly.
// ══════════════════════════════════════════════════════════════════════════════

// ── Lookup / Master DTOs ────────────────────────────────────────────────────

export interface LookupDTO {
  id: number;
  name: string;
  active: boolean;
  sort_order: number;
}

export interface StateDTO extends LookupDTO {
  NationalityID?: number | null;
}

export interface DistrictDTO extends LookupDTO {
  StateID: number;
}

export interface CourtDTO extends LookupDTO {
  DistrictID: number;
  StateID: number;
}

export interface UnitDTO extends LookupDTO {
  TypeID: number;
  ParentUnit?: number | null;
  NationalityID?: number | null;
  StateID: number;
  DistrictID: number;
}

export interface RankDTO extends LookupDTO {
  Hierarchy?: number | null;
}

export interface DesignationDTO extends LookupDTO {
  SortOrder?: number | null;
}

export interface EmployeeDTO {
  id: number;
  DistrictID: number;
  UnitID: number;
  RankID: number;
  DesignationID: number;
  KGID: string;
  FirstName: string;
  EmployeeDOB?: string | null;
  GenderID: number;
  BloodGroupID?: number | null;
  PhysicallyChallenged: boolean;
  AppointmentDate?: string | null;
  active: boolean;
}

export type CaseCategoryDTO = LookupDTO;
export type GravityOffenceDTO = LookupDTO;
export type CaseStatusDTO = LookupDTO;
export type CasteDTO = LookupDTO;
export type ReligionDTO = LookupDTO;
export type OccupationDTO = LookupDTO;
export type GenderDTO = LookupDTO;
export type NationalityDTO = LookupDTO;
export type BloodGroupDTO = LookupDTO;

export interface ActDTO {
  ActCode: string;
  ActDescription?: string | null;
  ShortName?: string | null;
  active: boolean;
}

export interface SectionDTO {
  ActCode: string;
  SectionCode: string;
  SectionDescription?: string | null;
  active: boolean;
}

export interface CrimeHeadDTO {
  id: number;
  CrimeGroupName: string;
  active: boolean;
}

export interface CrimeSubHeadDTO {
  id: number;
  CrimeHeadID: number;
  CrimeHeadName: string;
  SeqID?: number | null;
}

// ── Transactional DTOs — Create Payloads ────────────────────────────────────

export interface InvOccuranceTimeCreate {
  IncidentFromDate?: string | null;
  IncidentToDate?: string | null;
  InfoReceivedPSDate?: string | null;
  latitude?: number | null;
  longitude?: number | null;
  BriefFacts?: string | null;
}

export interface ComplainantDetailsCreate {
  ComplainantName: string;
  AgeYear?: number | null;
  OccupationID?: number | null;
  ReligionID?: number | null;
  CasteID?: number | null;
  GenderID?: number | null;
}

export interface FIRVictimCreate {
  VictimName: string;
  AgeYear?: number | null;
  GenderID?: number | null;
  VictimPolice: string;
}

export interface AccusedCreate {
  AccusedName: string;
  AgeYear?: number | null;
  GenderID?: number | null;
  PersonID?: string | null;
}

export interface ActSectionAssociationCreate {
  ActCode: string;
  SectionCode: string;
  ActOrderID?: number | null;
  SectionOrderID?: number | null;
}

export interface ArrestSurrenderCreate {
  ArrestSurrenderTypeID: number;
  ArrestSurrenderDate: string;
  ArrestSurrenderStateId: number;
  ArrestSurrenderDistrictId: number;
  PoliceStationID: number;
  IOID: number;
  CourtID: number;
  AccusedMasterID: number;
  IsAccused: boolean;
  IsComplainantAccused: boolean;
  associated_accused_ids: number[];
}

export interface ChargesheetDetailsCreate {
  csdate: string;
  cstype: string;
  PolicePersonID: number;
}

export interface CaseMasterCreate {
  CrimeNo?: string | null;
  CaseNo?: string | null;
  CrimeRegisteredDate?: string | null;
  PolicePersonID: number;
  PoliceStationID: number;
  CaseCategoryID: number;
  GravityOffenceID: number;
  CrimeMajorHeadID: number;
  CrimeMinorHeadID: number;
  CaseStatusID: number;
  CourtID: number;
  BriefFacts?: string | null;
  dataset_id?: number | null;
  occurrence_time?: InvOccuranceTimeCreate | null;
  complainants: ComplainantDetailsCreate[];
  victims: FIRVictimCreate[];
  accused: AccusedCreate[];
  act_sections: ActSectionAssociationCreate[];
  arrest_surrenders: ArrestSurrenderCreate[];
  chargesheets: ChargesheetDetailsCreate[];
}

// ── Transactional DTOs — Response Payloads ──────────────────────────────────

export interface InvOccuranceTimeResponse {
  CaseMasterID: number;
  IncidentFromDate?: string | null;
  IncidentToDate?: string | null;
  InfoReceivedPSDate?: string | null;
  latitude?: number | null;
  longitude?: number | null;
  BriefFacts?: string | null;
  created_at: string;
  updated_at: string;
}

export interface ComplainantDetailsResponse {
  id: number;
  CaseMasterID: number;
  ComplainantName: string;
  AgeYear?: number | null;
  OccupationID?: number | null;
  ReligionID?: number | null;
  CasteID?: number | null;
  GenderID?: number | null;
  created_at: string;
  updated_at: string;
}

export interface FIRVictimResponse {
  id: number;
  CaseMasterID: number;
  VictimName: string;
  AgeYear?: number | null;
  GenderID?: number | null;
  VictimPolice: string;
  created_at: string;
  updated_at: string;
}

export interface AccusedResponse {
  id: number;
  CaseMasterID: number;
  AccusedName: string;
  AgeYear?: number | null;
  GenderID?: number | null;
  PersonID?: string | null;
  created_at: string;
  updated_at: string;
}

export interface ActSectionAssociationResponse {
  CaseMasterID: number;
  ActCode: string;
  SectionCode: string;
  ActOrderID?: number | null;
  SectionOrderID?: number | null;
  created_at: string;
  updated_at: string;
}

export interface ArrestSurrenderResponse {
  id: number;
  CaseMasterID: number;
  ArrestSurrenderTypeID: number;
  ArrestSurrenderDate: string;
  ArrestSurrenderStateId: number;
  ArrestSurrenderDistrictId: number;
  PoliceStationID: number;
  IOID: number;
  CourtID: number;
  AccusedMasterID: number;
  IsAccused: boolean;
  IsComplainantAccused: boolean;
  associated_accused_ids: number[];
  created_at: string;
  updated_at: string;
}

export interface ChargesheetDetailsResponse {
  id: number;
  CaseMasterID: number;
  csdate: string;
  cstype: string;
  PolicePersonID: number;
  created_at: string;
  updated_at: string;
}

export interface CaseMasterResponse {
  id: number;
  CrimeNo?: string | null;
  CaseNo?: string | null;
  CrimeRegisteredDate?: string | null;
  PolicePersonID: number;
  PoliceStationID: number;
  CaseCategoryID: number;
  GravityOffenceID: number;
  CrimeMajorHeadID: number;
  CrimeMinorHeadID: number;
  CaseStatusID: number;
  CourtID: number;
  BriefFacts?: string | null;
  dataset_id?: number | null;
  created_at: string;
  updated_at: string;
  occurrence_time?: InvOccuranceTimeResponse | null;
  complainants: ComplainantDetailsResponse[];
  victims: FIRVictimResponse[];
  accused: AccusedResponse[];
  act_sections: ActSectionAssociationResponse[];
  arrest_surrenders: ArrestSurrenderResponse[];
  chargesheets: ChargesheetDetailsResponse[];
}

// ── Paginated List Response ─────────────────────────────────────────────────

export interface CaseListResponse {
  records: CaseMasterResponse[];
  total: number;
  page: number;
  page_size: number;
}

// ── All lookups combined for caching ────────────────────────────────────────

export interface FirLookupData {
  states: StateDTO[];
  ranks: RankDTO[];
  designations: DesignationDTO[];
  caseCategories: CaseCategoryDTO[];
  gravityOffences: GravityOffenceDTO[];
  caseStatuses: CaseStatusDTO[];
  castes: CasteDTO[];
  religions: ReligionDTO[];
  occupations: OccupationDTO[];
  genders: GenderDTO[];
  nationalities: NationalityDTO[];
  bloodGroups: BloodGroupDTO[];
  acts: ActDTO[];
  crimeHeads: CrimeHeadDTO[];
}
